### PR TITLE
add sleeping and spinning benchmark

### PR DESCRIPTION
**Pull request by HermioneKT**
This PR add 2 new benchmarks sleeping (to simulate I/O bounded task) and spinning (to simulate CPU bounded task) such that i can investigate how dynamically configure cpu frequency affects latency of such tasks

### CODE REVIEW

**Review by dhschall — COMMENTED** [review]
Hi @HermioneKT,
first of all thank you very much for contributing new benchmarks. Overall it looks already quite good but there are a couple of things to address.

First, both functions are almost the same. How about if you combine both, make one function and make it configurable with a input argument.
```
function = flag.String("functionality", "sleep", "Define the functionality of the function. Possible values: [sleep, spin]")
``` 
This way less files need to be maintained. Also, one can easily add a memory-intense version that randomly reads and writes from and to a large memory region (Btw. would be nice if you can add that already ;))
You can keep two yaml files. One with `- --functionality=sleep` and another one with `spin`.

Second, I think you should make the spinning as a delay as well. I don't know how much time it takes for the 30000 loop iterations. Is it long or short? Having a parameter that lets you spin for 20 ms is more intuitive.

Then we need to integrate it into the CI. Just copy the .github/workflows/e2e-gptj.yml file and replace all gptj references with your new image. I.e. https://github.com/vhive-serverless/vSwarm/blob/c5ae2f26d0456f582d108bd9207f91f4bd95f949/.github/workflows/e2e-gptj.yml#L38
No worries I'll help you with that :)

Small changes see my other comments.

Finally, before we merge you need to squash all commits into one commit and you also need to sign-off the commit. Refer to: https://github.com/vhive-serverless/vHive/blob/main/docs/contributing_to_vhive.md#commit-sign-off

Thanks, again and let me know if something is unclear or you need help

File: benchmarks/spinning/go/server.go
Comment (HermioneKT) [reply benchmarks/spinning/go/server.go:53]: Hi dhschall, first of all thank you so much for the review!! My apologies for these mistakes. I have no permission to push to vhiveeasse docker so i used my local, will asked for permission later. For these benchmarks i actually wanted to fix the iterations (instead of fixing it to spin for xx milliseconds) in order to check how latency changes when cpu frequency changes as part of my project to optimise power usage (tune down frequency for tasks like I/O bounded to save power since latency would not be affected by cpu frequency regardless) so we probably should not combine these 2 benchmarks together and make it run for a fixed amount of time, right? what do u think?

**Comment by dhschall** [comment]
I also added a CI workflow and tested it in #852.
Can you copy the `spin-sleep.yaml` in your code.
Also as you see here https://github.com/vhive-serverless/vSwarm/actions/runs/7856433672/job/21439224624#step:9:1 it fails to deploy in knative

**Comment by HermioneKT** [comment]
Hi @dhschall, I have fixed the end-to-end test, thank you! I would like to request review from you again if possible.

File: .github/workflows/e2e-spin-sleep.yml
Comment (dhschall) [inline .github/workflows/e2e-spin-sleep.yml:38]:
```suggestion
            sleeping-go,
            spinning-go
```

Comment (dhschall) [reply .github/workflows/e2e-spin-sleep.yml:38]: Please add the spinning benchmark otherwise it will never be tested

You also need a second parameter for the docker file:
Similar to:
```
matrix:
        include:
          - service: sleeping-go
            file: benchmarks/sleeping/docker/Dockerfile
```

File: .github/workflows/e2e-spin-sleep.yml
Comment (dhschall) [inline .github/workflows/e2e-spin-sleep.yml:71]:
```suggestion
```

The target is not needed since you only have one function per docker file

File: .github/workflows/e2e-spin-sleep.yml
Comment (dhschall) [inline .github/workflows/e2e-spin-sleep.yml:69]:
```suggestion
          file: ${{ matrix.file }}
```
Since you have different docker files you have to make this as a parameter.
Similar as you did for knative

**Review by dhschall — CHANGES_REQUESTED** [review]
Thank you @HermioneKT for the update. It looks all very great!
Great you made the CI work. Sometimes that is a bit annoying ;)

Can you now just add the second benchmark into the same file. Otherwise the second benchmark will never be build and tested.

File: benchmarks/sleeping/yamls/knative/kn-sleeping-go.yaml
Comment (HermioneKT) [reply benchmarks/sleeping/yamls/knative/kn-sleeping-go.yaml]: it is to control which node the benchmark should be invoked on. This is because one of my experiments requires spinning benchmark to be invoked on a high frequency node, sleeping benchmark on a low frequency node (which is why we need them in two different knative service with different knative yaml file to specify different nodeSelector) It did not pass the CI previously, because we need to run a command to enable this feature in Knative Serving that allows us to specify a NodeSelector in the pod spec of our Knative service first before deploying the service.

**Review by dhschall — APPROVED** [review]
Nice, thank you for your contribution.
It looks good to me. 
No worries about the linters. They seem to be broken.

**Pull request merged**
